docs/proposals/sysctl.md

```diff
@@ -0,0 +1,333 @@
+
+To get a list of all parameters, you can run
+
+```
+$ sudo sysctl -a
+```
+
+A number of them are namespaced and can therefore be set for a container
+independently with today's Linux kernels.
+
+**Note**: This proposal - while sharing some use-cases - does not cover ulimits
+(compare [Expose or utilize docker's rlimit support](https://github.com/kubernetes/kubernetes/issues/3595)).
+
+## Motivation
+
+A number of Linux applications need certain kernel parameter settings to
+
+- either run at all
+- or perform well.
+
+In Kubernetes we want to allow to set these parameters within a pod specification
+in order to enable the use of the platform for those applications.
+
+With Docker version 1.11.1 it is possible to change kernel parameters inside privileged containers.
+However, the process is purely manual and the changes might be applied across all containers
+affecting the entire host system. It is not possible to set the parameters within a non-privileged
+container.
+
+With [docker#19265](https://github.com/docker/docker/pull/19265) docker-run as of 1.12.0
+supports setting a number of whitelisted sysctls during the container creation process.
+
+Some real-world examples for the use of sysctls:
+
+- PostgreSQL requires `kernel.shmmax` and `kernel.shmall` (among others) to be
+  set to reasonable high values (compare [PostgresSQL Manual 17.4.1. Shared Memory
+  and Semaphores](http://www.postgresql.org/docs/9.1/static/kernel-resources.html)).
+  The default of 32 MB for shared memory is not reasonable for a database.
+- RabbitMQ proposes a number of sysctl settings to optimize networking: https://www.rabbitmq.com/networking.html.
+- web applications with many concurrent connections require high values for
+  `net.core.somaxconn`.
+- certain Java applications require "hugepages" support to perform well
+  (compare [docker#4717](https://github.com/docker/docker/issues/4717#issuecomment-77426026)),
+  configured through `vm.nr_hugepages`.
+- a containerized IPv6 routing daemon requires e.g. `/proc/sys/net/ipv6/conf/all/forwarding` and
+  `/proc/sys/net/ipv6/conf/all/accept_redirects` (compare
+  [docker#4717](https://github.com/docker/docker/issues/4717#issuecomment-98653017)).
+
+## Constraints and Assumptions
+
+* Only namespaced kernel parameters can be modified
+* Built on-top of the existing security context work
+* Be container-runtime agnostic
+  - on the API level
+  - the implementation (and the set of supported sysctls) will depend on the runtime
+* Kernel parameters can be set during a container creation process only.
+
+## Further work (out of scope for this proposal)
+
+* Update kernel parameters in running containers.
+* Integration with new container runtime proposal: https://github.com/kubernetes/kubernetes/pull/25899.
+
+## Use Cases
+
+As an administrator I want to set customizable kernel parameters for a container
+
+1. To be able to limit consumed kernel resources
+   1. so I can provide more resources to other containers
```

Comment:
In general I agree with starting with the node first. I feel that it will help figuring out how we would address scheduling as part of this proposal even if we don't implement it right away. That way, we are sure that scheduler integration is possible and that there will not be a need to alter node or API behavior.